Limiting pagination buttons

elboletaire opened the pull request:
TODOs (mental notes):
- ~~Remap how urls work so we don't have page 0, despite the API requiring it...~~
- ~~Ensure ellipsis buttons are considered into the `maxButtons` property count~~
- ~~Take in mind what happens if no totalPages is defined~~
- ~~improve navigation to ellipsed pages~~
- Move this components to @vocdoni/chakra-components, since we have the code duplicated here and on ui-scaffold.

elboletaire committed:
Limited pagination using `maxButton` prop

- Also made it fully customizable via direct props for the ButtonGroup,
  or via buttonProps for all the page button props
- Now we don't use `0` in the URL for paginating results, despite the
  API requiring it

I'd still move this to ui-components at some point, because we're gonna
need it in ui-scaffold sooner than later.

selankon commented on src/components/Pagination/PaginationProvider.tsx:
```diff
@@ -3,7 +3,7 @@ import { createContext, PropsWithChildren, useContext, useState } from 'react'
   page: number
   setPage: (page: number) => void
-  totalPages: number
+  totalPages?: number
```
If total pages is undefined and next/previous button are shown, could be easy to add a "growing up" pagination buttons. As you know the page you actually are, it could show just the button of the actual  page and the previous one (using elipsis if `maxButtons` is defined).

Similar at how it works now on the actual explorer. For example searching by elections with results on stg explorer you will see the behavior. 

Is not needed but we can consider to implement it

elboletaire replied:
Ets un cabró.

selankon commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,45 +1,199 @@
+          </Button>
+        </>
+      ) : (
+        pages
```
This should be wrapped on a flex wrap or similar for little screens

![image](https://github.com/vocdoni/explorer/assets/16777076/43cf9060-c6c3-4993-949a-e7e66790e458)

elboletaire replied:
The component can receive props for its button group and for the buttons independently.

I'm not styling it (at least not yet), since I first need to get it out of here and move it to chakra-components... and the styles are something generic for the app.

selankon commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,45 +1,199 @@
+      return pages
+    }
+
+    const startEllipsis = createEllipsisButton(
+      'start-ellipsis',
+      (e) => {
+        e.preventDefault()
+        setStartEllipsisInput(true)
+      },
+      startEllipsisInput,
+      (e) => {
+        if (e.key === 'Enter') {
+          const pageNumber = Number((e.target as HTMLInputElement).value)
+          gotoPage(pageNumber - 1)
+          setStartEllipsisInput(false)
+        }
+      },
+      () => setStartEllipsisInput(false)
+    )
+
+    const endEllipsis = createEllipsisButton(
+      'end-ellipsis',
+      (e) => {
+        e.preventDefault()
+        setEndEllipsisInput(true)
+      },
+      endEllipsisInput,
+      (e) => {
+        if (e.key === 'Enter') {
+          const pageNumber = Number((e.target as HTMLInputElement).value)
+          gotoPage(pageNumber - 1)
+          setEndEllipsisInput(false)
+        }
+      },
+      () => setEndEllipsisInput(false)
+    )
```
This code is duplicated. Why just don't store the ellipsis input state inside the component and pass the `gotoPage` as parameter of the component?

```ts
const EllipsisButton = ({ key, gotoPage }: { key: string; gotoPage: (page: number) => void }) => {
  const [ellipsisInput, setEllipsisInput] = useState(false)

  if (ellipsisInput) {
    return (
      <Input
        key={key}
        placeholder='Page #'
        width='50px'
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            const pageNumber = Number((e.target as HTMLInputElement).value)
            gotoPage(pageNumber - 1)
            setEllipsisInput(false)
          }
        }}
        onBlur={() => setEllipsisInput(false)}
        autoFocus
      />
    )
  } else {
    return (
      <Button
        key={key}
        as='a'
        href='#goto-page'
        onClick={(e) => {
          e.preventDefault()
          setEllipsisInput(true)
        }}
      >
        ...
      </Button>
    )
  }
}
```

And then just:

```ts
    const startEllipsis = <EllipsisButton key={'start-ellipsis'} gotoPage={gotoPage} />
    const endEllipsis = <EllipsisButton key={'end-ellipsis'} gotoPage={gotoPage} />
```

Or similar. Provably you could use just use one `EllipsisButton` component

elboletaire replied:
True, it ended up this way after lots of iterations of the code. But I'll change it.

selankon commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,45 +1,199 @@
+    const sideButtons = 2 // First and last page
+    const availableButtons = maxButtons - sideButtons // Buttons we can distribute around the current page
+
+    if (currentPage <= availableButtons / 2) {
+      // Near the start
+      return [...pages.slice(0, availableButtons), endEllipsis, pages[totalPages - 1]]
+    } else if (currentPage >= totalPages - 1 - availableButtons / 2) {
+      // Near the end
+      return [pages[0], startEllipsis, ...pages.slice(totalPages - availableButtons, totalPages)]
+    } else {
+      // In the middle
+      const startPage = currentPage - Math.floor((availableButtons - 1) / 2)
+      const endPage = currentPage + Math.floor(availableButtons / 2)
+      return [pages[0], startEllipsis, ...pages.slice(startPage + 1, endPage), endEllipsis, pages[totalPages - 1]]
+    }
```
There are a strange behavior if the max buttons is set to `1`... Shall we handle it? For example, not showing the Ellipsis buttons or  either show the page button with the previous/next buttons? 

![image](https://github.com/vocdoni/explorer/assets/16777076/4dcca5b4-dcf9-4458-b91a-a806944f937e)

elboletaire replied:
Not sure we should handle the "1" max buttons. Why do you want a single button for pagination? 

Edit: same for numbers below 4 in general I think... The minimum logic could be 3 (first page, ellipsis, last page), but then how would I manage the current page when outside the first or last ones? I'd be showing 5 buttons already. So I think this should be hard-limited to 5

selankon reviewed (commented):
Is a nice work and the code seems really much scalable than the actual explorer paginator. However there are some things to take a look. 

Another thing I miss from the actual explorer paginator is a previous/next button along with the first/last page buttons. Not required but can be a nice to have feature:

![image](https://github.com/vocdoni/explorer/assets/16777076/095262fb-1089-4e37-908b-55978e4b1a8d)

selankon commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,45 +1,199 @@
+  onKeyDown: (e: React.KeyboardEvent<HTMLInputElement>) => void,
+  onBlur: () => void
+) => {
+  if (showInput) {
+    return <Input key={key} placeholder='Page #' width='50px' onKeyDown={onKeyDown} onBlur={onBlur} autoFocus />
+  } else {
+    return (
+      <Button key={key} as='a' href='#goto-page' onClick={onClick}>
+        ...
+      </Button>
+    )
+  }
+}
```
This else can be deleted

```suggestion
  if (showInput) {
    return <Input key={key} placeholder='Page #' width='50px' onKeyDown={onKeyDown} onBlur={onBlur} autoFocus />
  }
  return (
    <Button key={key} as='a' href='#goto-page' onClick={onClick}>
      ...
    </Button>
  )
}
```

elboletaire replied:
💯 Agree

selankon reviewed (commented):
I forgot this

elboletaire committed:
Remove unnecessary else

elboletaire committed:
Simplify ellipsis creation with a custom component

Also fixed some issues related to old pagination flow (removing
unnecessary adds and substracts).

And passed missing arguments to both ellipsis buttons and ellipsis
inputs.

elboletaire committed:
Hard limit min number of maxButtons to 5

elboletaire commented:
> Another thing I miss from the actual explorer paginator is a previous/next button along with the first/last page buttons. Not required but can be a nice to have feature:

I have it in mind. Let's add it if someone's requests it.

elboletaire committed:
Fix keys and pages issue

elboletaire committed:
Fix pages position after latest changes
